### DISCUSSION TITLE
How can I integrate next-auth with Nestjs

### DISCUSSION THREAD

**Discussion by cloudcompute** [nextauthjs/next-auth#1775] [Help] [answered]
Next is a framework that serves the React front end to the user. While, on the other hand, auth functionality is something that is implemented on the backend and it uses a db to store the Sessions data. How next-auth (kind of backend functionality) works in conjunction with the Next (frontend framework)?

My Nest JS based application exposes the GraphQL API to the client apps. How can I integrate these three things i.e. next-auth, Next, and Nest while making sure that the calls to the Prisma orm / db should only be made from the NestJS app.

Kindly guide. Thanks

**Comment by StefanSelfTaught** [comment]
NextAuth is making use of [Next API routes](https://nextjs.org/docs/api-routes/introduction) which is Next solution for providing a minimal backend and each 'api route' can be deployed for example as a serverless function. 

The database with NextAuth is [optional](https://next-auth.js.org/faq#should-i-use-a-database) if you don't need to persist user data or support email sign in. If you don't specify a database then JWTs will be enabled for session storage and used to store session data.

It will be a problem for you if  both Nest and Next will interact with your database?

**Comment by j-lee8** [comment]
I had a similar issue as this. I wanted to use a separate Koa backend but there are no examples anywhere lol. All I know is you need to use the authorise/jwt callback to fetch a token from the server and store it in your Next session.

Then when you make calls to your Nest server, you need to check this token (that you pass in from Next) to determine whether the user is authorised. In Apollo server I think you need to update the context to do this, something like:

`const user = getToken(ctx.headers.authorization)`

Unfortunately, as I said, there are no end-to-end examples anywhere. I know people post articles and posts with brevity in mind, but it's incredibly frustrating because connecting the dots isn't always straightforward. Some times you just need a codebase to look at to learn.

I just went for API routes in the end with a GraphQL endpoint at `api/graphql`. It's not what I wanted, but it'll do for a MVP.

**Reply by cloudcompute** [reply]
Very true.. when you said "people post articles and posts with brevity in mind" It is probably because they assume that all the readers of the article have exactly the similar amount of knowledge that they have for a given topic. They usually just write the portion of a topic that they learnt recently.

**Comment by cloudcompute** [comment]
Hi

@StefanSelfTaught @JavaJamie 

The following repository is using both nestjs and next-auth
https://github.com/tn/next-nest-boilerplate/blob/fdb51f4ec9c78f9c2e3d3ae7e446a7bccd649ba5/server/guards/auth.guard.ts#L2

Do you think that is this the right way to use it?

Thanks

**Reply by StefanSelfTaught** [reply]
@balazsorban44

**Reply by balazsorban44** [reply]
I don't have enough knowledge of Nest to really understand what is happening, but `getSession` will try to make a fetch call to the `/api/auth/session` endpoint. If it is set up in that context, then I guess it should work.
Unfortunately I don't have the capacity to check this out myself.

**Reply by daniloraisi** [reply]
> I don't have enough knowledge of Nest to really understand what is happening, but `getSession` will try to make a fetch call to the `/api/auth/session` endpoint. If it is set up in that context, then I guess it should work.
> Unfortunately I don't have the capacity to check this out myself.

I'm using Nest practically on a daily basis, and analyzing the code from this repo it should work fine! Considering using this as a starting point for my next project.

**Comment by balazsorban44** [comment, marked as answer]
There is a discussion in #2294 to expand to other frameworks than Next.js. In the short(er) term, we could aim for anything with a backend and frontend using React. in long(er) terms, we could expand to Svelte, Vue, etc.

**Reply by vittarr** [reply]
Can you please advise when you have plans to extend to Vue /Nuxt

**Comment by raychunghk** [comment]
I asked AI  and it already help half of the work.
https://poe.com/s/F88S8OyXYkv7pEmJRAga

**Reply by alejandroch1202** [reply]
Interesting

**Reply by luke358** [reply]
but,when your next-auth session is expires, it have lots of error. Or when i refresh token, but next-auth session not work
